crates/subspace-farmer/src/bin/subspace-farmer/commands/bench.rs

```diff
@@ -306,3 +306,202 @@ fn get_size(path: impl AsRef<Path>) -> std::io::Result<u64> {
+        let path = seeder_base_directory.as_ref().join("object-mappings");
+
+        move || ObjectMappings::open_or_create(path)
+    })
+    .await
+    .unwrap()
+    .unwrap();
+
+    let base_path = seeder_base_directory.as_ref().to_owned();
+    let plot_factory = move |plot_index, public_key, max_piece_count| {
+        let base_path = base_path.join(format!("plot{plot_index}"));
+        Plot::open_or_create(base_path, public_key, max_piece_count)
+    };
+
+    let seeder_multi_farming = MultiFarming::new(
+        MultiFarmingOptions {
+            base_directory: seeder_base_directory.as_ref().to_owned(),
+            archiving_client: seeder_client.clone(),
+            farming_client: seeder_client.clone(),
+            object_mappings: object_mappings.clone(),
+            reward_address: PublicKey::default(),
+            bootstrap_nodes: vec![],
+            listen_on: vec![seeder_multiaddr.clone()],
+            dsn_sync: false,
+        },
+        seeder_max_plot_size,
+        seeder_max_plot_size,
+        plot_factory,
+        false,
+    )
+    .await
+    .unwrap();
+
+    {
+        let mut last_archived_block = LastArchivedBlock {
+            number: 0,
+            archived_progress: ArchivedBlockProgress::Partial(0),
+        };
+
+        for segment_index in 0..seeder_max_plot_size / PIECE_SIZE as u64 / 256 {
+            last_archived_block
+                .archived_progress
+                .set_partial(segment_index as u32 * 256 * PIECE_SIZE as u32);
+
+            let archived_segment = {
+                let root_block = RootBlock::V0 {
+                    segment_index,
+                    records_root: Sha256Hash::default(),
+                    prev_root_block_hash: Sha256Hash::default(),
+                    last_archived_block,
+                };
+
+                let mut pieces = FlatPieces::new(256);
+                rand::thread_rng().fill(pieces.as_mut());
+
+                let objects = std::iter::repeat_with(|| PieceObject::V0 {
+                    hash: rand::random(),
+                    offset: rand::random(),
+                })
+                .take(100)
+                .collect();
+
+                ArchivedSegment {
+                    root_block,
+                    pieces,
+                    object_mapping: vec![PieceObjectMapping { objects }],
+                }
+            };
+
+            if seeder_archived_segments_sender
+                .send(archived_segment)
+                .await
+                .is_err()
+            {
+                break;
+            }
+        }
+    }
+    seeder_multiaddr.push({
+        schnorrkel::PublicKey::from_bytes(
+            &*seeder_multi_farming.single_plot_farms[0].plot.public_key(),
+        )
+        .map(identity::sr25519::PublicKey::from)
+        .map(identity::PublicKey::Sr25519)
+        .map(PeerId::from)
+        .map(Into::into)
+        .map(Protocol::P2p)
+        .unwrap()
+    });
+
+    for _ in 0..10 {
```

Comment:
10 times because it is probabilistic? Can we use hand-picked IDs instead and verify scenarios that we are interested in?

Reply:
Hmm, maybe we can check in advance how many pieces should be retrieved for some identity and just check for it?

Reply:
Hmm, I see why it is bad to have random in tests (especially in this kind of ones), but I think that having seeded tests is unhelpful. We generally want to test that some property holds for any input, so if it fails, we want to fix it.

I would just save here all the piece index hashes into btree map and just compare number of pieces which were synced and piece range.

Reply:
I understand, but for that you analyze edge-cases and write tests that test both sides of an edge case to ensure correct behavior. If you want to find unknown errors you're going into fuzzing territory. With randomized test you'll not test most of the edge-cases most of the time, which is why it is suboptimal.